components/supervisor/pkg/supervisor/services.go

```diff
@@ -1114,70 +1119,54 @@ func (s *taskService) ListenToOutput(req *api.ListenToOutputRequest, srv api.Tas
 					return
 				}
 				for _, update := range updates {
-					if update == nil {
-						continue
-					}
-					if update.Id == req.TaskId && update.State == api.TaskState_closed {
+					if update != nil && update.Id == req.TaskId && update.State == api.TaskState_closed {
 						return
 					}
 				}
 			}
 		}
 	}()
 
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return status.Error(codes.Internal, err.Error())
-	}
-	defer watcher.Close()
-
-	dir, fileName := filepath.Split(fileLocation)
-
-	err = watcher.Add(dir)
-	if err != nil {
-		return status.Error(codes.Internal, err.Error())
-	}
-
-	// Send all previous logs before starting to watch for new ones
 	offset := int64(0)
-	for {
-		n, err := file.ReadAt(buf, offset)
-		if n > 0 {
-			offset += int64(n)
-			if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
+
+	readAndStream := func() error {
+		for {
+			n, err := file.ReadAt(buf, offset)
+			if n > 0 {
+				offset += int64(n)
+				if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
+					return status.Error(codes.Internal, err.Error())
+				}
+			}
+			if err == io.EOF {
+				return nil
+			}
+			if err != nil {
 				return status.Error(codes.Internal, err.Error())
 			}
 		}
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return status.Error(codes.Internal, err.Error())
-		}
 	}
 
 	for {
 		select {
 		case <-done:
-			return nil
+			// Task closed or context done; exit after sending remaining logs
+			return readAndStream()
 		case event := <-watcher.Events:
-			if event.Op&fsnotify.Write == fsnotify.Write && filepath.Base(event.Name) == fileName {
-				n, err := file.ReadAt(buf, offset)
-				if n > 0 {
-					offset += int64(n)
-					if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
-						return status.Error(codes.Internal, err.Error())
-					}
-				}
-				if err != nil && err != io.EOF {
-					return status.Error(codes.Internal, err.Error())
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				if err := readAndStream(); err != nil {
+					return err
 				}
 			}
 		case err := <-watcher.Errors:
 			if err != nil {
 				log.Println("error:", err)
 				return status.Error(codes.Internal, err.Error())
 			}
+		default:
+			if err := readAndStream(); err != nil {
```

Comment:
Can also be `return readAndStream()`

Reply:
Wouldn't that just jump out of the loop, ending the stream even if the task was still running?